### PR TITLE
New lint: `disallowed_script_idents`

### PR DESCRIPTION
**Pull request by popzxc**
This PR implements a new lint to restrict locales that can be used in the code,
as proposed in #7376.

Current concerns / unresolved questions:

- ~~Mixed usage of `script` (as a Unicode term) and `locale` (as something that is easier to understand for the broad audience). I'm not sure whether these terms are fully interchangeable and whether in the current form it is more confusing than helpful.~~ `script` is now used everywhere.
- ~~Having to mostly copy-paste `AllowedScript`. Probably it's not a big problem, as the list of scripts is standardized and is unlikely to change, and even if we'd stick to the `unicode_script::Script`, we'll still have to implement custom deserialization, and I don't think that it will be shorter in terms of the amount of LoC.~~ `unicode::Script` is used together with a filtering deserialize function.
- Should we stick to the list of "recommended scripts" from [UAX #31](http://www.unicode.org/reports/tr31/#Table_Recommended_Scripts) in the configuration? 

*Please write a short comment explaining your change (or "none" for internal only changes)*

changelog: ``[`disallowed_script_idents`]``

r? @Manishearth

### CODE REVIEW

File: clippy_lints/src/restricted_locales.rs
Comment (Manishearth) [inline clippy_lints/src/restricted_locales.rs:40]: restriction

File: clippy_lints/src/restricted_locales.rs
Comment (Manishearth) [inline clippy_lints/src/restricted_locales.rs:74]: script, not locale

Comment (Manishearth) [reply clippy_lints/src/restricted_locales.rs:74]: also, "not allowed by configuration"

File: clippy_lints/src/restricted_locales.rs
Comment (Manishearth) [inline clippy_lints/src/restricted_locales.rs:89]: Kinda feel like I'd rather use the same enum and filter out these variants when the lint object is constructed

File: clippy_lints/src/restricted_locales.rs
Comment (Manishearth) [inline clippy_lints/src/restricted_locales.rs:58]: Given that lint passes are always run, can you check if there's a measurable perf difference when linting a decently sized crate?

File: clippy_lints/src/utils/conf.rs
Comment (Manishearth) [inline clippy_lints/src/utils/conf.rs:215]: scripts, not locales

File: clippy_lints/src/restricted_locales.rs
Comment (Manishearth) [inline clippy_lints/src/restricted_locales.rs:39]: scripts, not locales

File: clippy_lints/src/restricted_locales.rs
Comment (Manishearth) [inline clippy_lints/src/restricted_locales.rs:22]: scripts, not locales

**Review by Manishearth — COMMENTED** [review]
Might be good to get a second set of eyes on this, cc @camsteffen

File: clippy_lints/src/restricted_locales.rs
Comment (popzxc) [reply clippy_lints/src/restricted_locales.rs:58]: It seems that I cannot just build the release version of clippy and run it. `clippy-driver` fails with the following error: `dyld: Library not loaded: @rpath/librustc_driver-22f067a38d2152ee.dylib`.

If I understood things correctly from looking through this repo issues, I need the whole rust toolchain for that, which is not desirable. Is there a simpler way to run clippy on a custom crate?

**Comment by camsteffen** [comment]
This lint is analogous to [`non_ascii_idents`](https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#non-ascii-idents). I think we should add "See also: [`non_ascii_idents`](https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#non-ascii-idents)" to the docs. And maybe the lint name should have some similarity, like `disallowed_script_idents`.

**Comment by Manishearth** [comment]
Yeah, I like that!

**Review by camsteffen — CHANGES_REQUESTED** [review]

File: clippy_lints/src/restricted_scripts.rs
Comment (Manishearth) [reply clippy_lints/src/restricted_scripts.rs:96]: Yes, if someone is willing to make the PR I'd merge it!

Comment (camsteffen) [reply clippy_lints/src/restricted_scripts.rs:96]: There is an existing function `Script::full_name` which seems to be to what the Unicode docs refer to as "alias". So would you name the new function `Script::from_full_name`?

Comment (popzxc) [reply clippy_lints/src/restricted_scripts.rs:96]: Would it make sense to add it now (so the new version will be released and used here), or for now it's better to implement conversion logic in `clippy` and replace it in one of the future releases?

File: clippy_lints/src/restricted_scripts.rs
Comment (popzxc) [reply clippy_lints/src/restricted_scripts.rs:75]: While the suggestion makes sense, I don't quite understand the following

> and checking that each `Script` is in `whitelist`.

IIUTC, `ScriptExtension` contains the list of scripts the character *may belong* to. Thus, wouldn't it be more right to check that *at least one* `Script` is in the `whitelist`? Like, "can this character be use in a context of whitelisted scripts?".

File: clippy_lints/src/restricted_scripts.rs
Comment (popzxc) [reply clippy_lints/src/restricted_scripts.rs:19]: Well, it looks like it'll be more complicated to implement. I can see, for example, the alias `Katakana_Or_Hiragana`, which represents *two* scripts rather than one.
In theory it could've implemented through `ScriptExtension`s, but I don't see a way to construct it manually, which means that we'll have to return `Vec<Script>`, which is less efficient.

Do you find it significant to stick to the aliases?

File: clippy_lints/src/restricted_scripts.rs
Comment (popzxc) [reply clippy_lints/src/restricted_scripts.rs:71]: Reimplemented using `check_crate`.

File: clippy_lints/src/restricted_scripts.rs
Comment (popzxc) [reply clippy_lints/src/restricted_scripts.rs]: For now, implemented as a check for at least one script. Will change if I just didn't understand something.

File: clippy_lints/src/restricted_scripts.rs
Comment (camsteffen) [reply clippy_lints/src/restricted_scripts.rs]: > Well, it looks like it'll be more complicated to implement. I can see, for example, the alias `Katakana_Or_Hiragana`, which represents _two_ scripts rather than one.
> In theory it could've implemented through `ScriptExtension`s, but I don't see a way to construct it manually, which means that we'll have to return `Vec<Script>`, which is less efficient.

Okay, this had me puzzled for a bit but I finally figured it out - it is in fact one script but it is no longer used in the latest version of Unicode. I think we can just ignore special cases like this. I looked at the source of the `unicode-script` crate and it is definitely using the "Alias" values.

> Do you find it significant to stick to the aliases?

Yes, it is better to use a pre-existing naming convention.

File: clippy_lints/src/restricted_scripts.rs
Comment (camsteffen) [reply clippy_lints/src/restricted_scripts.rs]: I think it would be good to add it to `unicode-script` now. Would you be interested in doing that?

File: clippy_lints/src/restricted_scripts.rs
Comment (camsteffen) [reply clippy_lints/src/restricted_scripts.rs]: You're exactly right! 👍

File: clippy_lints/src/restricted_scripts.rs
Comment (camsteffen) [reply clippy_lints/src/restricted_scripts.rs]: FYI I learned about Katakana_Or_Hiragana being a special case on [this page](https://unicode.org/reports/tr31/).

File: clippy_lints/src/restricted_scripts.rs
Comment (popzxc) [reply clippy_lints/src/restricted_scripts.rs]: Yeah, no problem, will take care of that tomorrow.

Comment (popzxc) [reply clippy_lints/src/restricted_scripts.rs]: [Here it is](https://github.com/unicode-rs/unicode-script/pull/7)

File: clippy_lints/src/restricted_locales.rs
Comment (Manishearth) [reply clippy_lints/src/restricted_locales.rs]: You need to have the rustc-dev component installed. You're probably using a different toolchain when you build clippy vs when you run it.

Check out https://github.com/rust-lang/rust-clippy/blob/master/rust-toolchain#L3

File: clippy_lints/src/restricted_scripts.rs
Comment (Manishearth) [reply clippy_lints/src/restricted_scripts.rs]: Oh, if you're looking for handling that special case you should use the AugmentedScriptSet property from unicode-security instead

File: clippy_lints/src/restricted_scripts.rs
Comment (Manishearth) [reply clippy_lints/src/restricted_scripts.rs]: And that code is published now

File: clippy_lints/src/restricted_scripts.rs
Comment (popzxc) [reply clippy_lints/src/restricted_scripts.rs]: I added a note with a link to the page about it being a special case instead. I think that it's better than adding one more dependency.

File: clippy_lints/src/restricted_locales.rs
Comment (popzxc) [reply clippy_lints/src/restricted_locales.rs]: Dunno. The compiler is (and was) overridden to be the same as for clippy itself. Components are also seem to be installed.

```sh
$ rustup +nightly-2021-06-17 component add llvm-tools-preview 
info: component 'llvm-tools-preview' for target 'aarch64-apple-darwin' is up to date
$ rustup +nightly-2021-06-17 component add rust-src          
info: component 'rust-src' is up to date
$ rustup +nightly-2021-06-17 component add rustc-dev         
info: component 'rustc-dev' for target 'aarch64-apple-darwin' is up to date
$ rustup show
# ...
nightly-2021-06-17-aarch64-apple-darwin (directory override for '.../actix-web')
# ...
```

Yet it still fails:

```sh
$ ../rust-clippy/target/release/clippy-driver --deny clippy::needless_borrow
dyld: Library not loaded: @rpath/librustc_driver-22f067a38d2152ee.dylib
  Referenced from: /<...>/actix-web/../rust-clippy/target/release/clippy-driver
  Reason: image not found
[1]    32340 abort      ../rust-clippy/target/release/clippy-driver --deny clippy::needless_borrow
```

If that's important, I build clippy via simple `cargo build --release` (maybe I need to do it some other way?).

File: clippy_lints/src/restricted_scripts.rs
Comment (camsteffen) [reply clippy_lints/src/restricted_scripts.rs]: There are lots of scripts that are not used in current Unicode. Katakana_Or_Hiragana is just one of them. I don't think it's anything we need to document or worry about.

Comment (popzxc) [reply clippy_lints/src/restricted_scripts.rs]: OK then, removed 😅

**Comment by popzxc** [comment]
So, apart from the perfing this lint (which I have some troubles with, unfortunately, so I'll be grateful if someone else have tried it), is there anything else I should improve?

**Review by Manishearth — APPROVED** [review]
I think this is good, and with the global bailing-out there's no need to test for perf

(Going to wait for a second review from @camsteffen)

File: clippy_lints/src/disallowed_script_idents.rs
Comment (camsteffen) [inline clippy_lints/src/disallowed_script_idents.rs:9]:
```suggestion
    /// **What it does:** Checks for usage of unicode scripts other than those explicitly allowed
```

**Review by camsteffen — APPROVED** [review]
Looks good to me!

**Comment by popzxc** [comment]
Uhm. It seems that at the same time I squashed commits and bors was r+ed, which caused it to be stuck.

I think `bors retry` is required now @Manishearth 😅

**Pull request merged**
